Don't configure ReferenceManager when doing form discovery parse

lognaturel opened the pull request:
Closes #4634 

#### What has been done to verify that this works as intended?
The goal is to make it so `ReferenceManager` is only configured when a user loads a form for filling. I did a search through the codebase for `ReferenceManager.` and made sure all calls are access-only. I removed configuration from the disk sync task and metadata parsing (incidentally, there was some redundancy there so this is a good change overall).

Allowing form parses when a secondary instance is required without a configured `ReferenceManager` required JavaRosa changes. Those are at https://github.com/getodk/javarosa/pull/635 and include some new tests. The JR snapshot has now been updated to include those changes.

Manually tried the following cases:
- Download a form that requires a CSV external secondary instance [external-select-csv.xml.zip](https://github.com/getodk/collect/files/6712840/external-select-csv.xml.zip) without attaching the CSV. The download succeeds. Opening the form succeeds but there are no choices available.
-  Download a form that requires a CSV external secondary instance used in a select and attach a CSV with missing `label` and `name` columns. The download fails with `Failure`, as it does on master.
- Download a form that requires a CSV external secondary instance without attaching the CSV. Then attach a CSV with missing `label` and `name` columns and download the update. The update succeeds. Empty choices are displayed. Selecting a choice and moving to the next question shows an error.
- Download a form in one project, switch projects to a project with a form that has media, verify that the media is visible (e.g. All Widgets and Likert widget images)
- Download a form with a last-saved reference, then open it and verify last-saved is used
- Verify forms with media work on first and second open and with switches between projects
- adb push forms that have a last-saved call, that require an external CSV without pushing that CSV, verify they are correctly discovered. Verify adding the external CSV makes the choices appear.

#### Why is this the best possible solution? Were any other approaches considered?
It's relatively simple and improves Collect code quality.

I tried to add a helpful message when the file is missing (https://github.com/getodk/collect/pull/4643/commits/b953ef946eb1767d7ba4e60184a31a289c19e9f5) but I backed out of that because we can't currently tell the difference between a missing external secondary instance and a choice filter that doesn't return any items. Adding a message made a common case worse: having a cascade of selects in a field-list. In that case, you want all the lower-level questions to show no options until selections are made further up the chain.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This is pretty risky. Everything that touches media is affected. Everything that discovers forms is affected (automatic form downloads, disk sync, manual form download). The biggest risk is that media doesn't show up somewhere it should, especially when switching projects.

#### Do we need any specific form for testing your changes? If so, please attach one.
See cases above. Forms with last-saved an XML and CSV external secondary instances.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

lognaturel committed:
Don't configure ReferenceManager for form discovery

Reverts https://github.com/getodk/collect/pull/3523 to avoid issues when form discovery happens in multiple projects at once or concurrent to form filling

lognaturel committed:
Remove ReferenceManager configuration for last-saved

lognaturel commented:
@seadowg we'd talked about me merging and doing another beta but since the last beta we published is still in review for unclear reasons I'd rather wait. I think it would also be a good idea for you to do a quick sanity check to try to identify any cases I may have still missed.

lognaturel commented on collect_app/src/test/java/org/odk/collect/android/formmanagement/FormMetadataParserTest.java:
```diff
@@ -60,43 +59,17 @@ public void canParseFormWithLastSaved() throws Exception {
         assertThat(metaData.get(FileUtils.FORMID), is("basic-last-saved"));
     }
 
-    /**
```
This will now parse with the placeholder instance but there will be no choices to pick from. That case is tested in JavaRosa. It feels esoteric enough that it doesn't need to be tested again.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/utilities/FormUtils.java:
```diff
@@ -28,17 +28,13 @@ private FormUtils() {
+        referenceManager.reset();
 
-        // This should get moved to the Application Class
-        if (referenceManager.getFactories().length == 0) {
```
When I was in here recently (fixing the `StoragePathProvider` call) I didn't think the check made sense at all - why would we care if there are already configured factories? I'm guessing our thinking lines up here, but I just wanted to check.

lognaturel replied:
It used to work out because there was only one factory to add -- the one on the ODK root. I think at that point without the check there would just have been lots of factories with the same root. Now we use a whole different context per project so I thought it was clearer to just always reset and configure a fresh factory (as opposed to detecting whether we've changed projects and only reconfiguring if we have).

There's a lot of weirdness here in general. I don't know why JR decided to have different prefixes for different file types. This whole code here is set up to erase that distinction -- we basically say no matter what the prefix is, first pretend it's jr://file and then resolve that.

lognaturel replied:
Oh and just another note since I was also confused by all this. There used to be a call on `referenceManager.clearSession()` which did NOT reset the factories. That's why the test had value and why without it there'd be lots of duplicate factories.

seadowg commented on collect_app/src/test/java/org/odk/collect/android/formmanagement/FormMetadataParserTest.java:
```diff
@@ -37,7 +36,7 @@ public void canParseFormWithExternalSecondaryInstance() throws Exception {
         File externalInstance = new File(mediaDir, "external-data.xml");
         FileUtils.write(externalInstance, EXTERNAL_INSTANCE.getBytes());
 
-        FormMetadataParser formMetadataParser = new FormMetadataParser(referenceManager);
```
I think we also want to remove the `referenceManager` field from the test class as well right? There are tests that use it that happen to pass right now but they're definitely misleading.

lognaturel committed:
Remove ReferenceManager-related tests

These still capture the intended end state but there's no interaction with ReferenceManager in form metadata parsing anymore so they make sense to remove. It's also the case that if a form was just opened and then is updated, that form's media files would resolve during and after the medata parse because the ReferenceManager is only reset at the beginning of a form filling session.
